pennylane/workflow/construct_batch.py

```diff
@@ -0,0 +1,268 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a function extracting the tapes at postprocessing at any stage of a transform program.
+
+"""
+import inspect
+from typing import Union, Callable, Tuple
+
+import pennylane as qml
+from .qnode import QNode, _make_execution_config, _get_device_shots
+
+
+def _get_full_transform_program(qnode: QNode) -> "qml.transforms.core.TransformProgram":
+    program = qml.transforms.core.TransformProgram(qnode.transform_program)
+    if getattr(qnode.gradient_fn, "expand_transform", False):
+        program.add_transform(
+            qml.transform(qnode.gradient_fn.expand_transform),
+            **qnode.gradient_kwargs,
+        )
+    if isinstance(qnode.device, qml.devices.Device):
+        config = _make_execution_config(qnode)
+        return program + qnode.device.preprocess(config)[0]
+    program.add_transform(qml.transform(qnode.device.batch_transform))
+    program.add_transform(qml.transforms.core.expand_fn_transform(qnode.device.expand_fn))
+    return program
+
+
+def transform_program(qnode: "QNode", level=None) -> "qml.transforms.core.TransformProgram":
+    """Extract a transform program at a designated level.
+
+    Args:
+        qnode (QNode): the qnode to get the transform program for.
+        level (None, str, int, slice): And indication
+
+            * ``None``: use the full transform program
+            * ``str``: Acceptable keys are ``"user"``, ``"device"``, and ``"gradient"``
+            * ``int``: How many transforms to include, starting from the front of the program
+            * ``slice``: a slice to select out components of the transform program.
```

Comment:
So yes, there are a lot of different ways we can designated the level, but I would call the main responsibility of this function is interpreting what the different levels mean.

So this will be the one source of truth for interpreting these designations.  `construct_batch` then relies on this function, and `draw`, `draw_mpl`, and `specs` will rely on `construct_batch`.

The most important level designations are the special strings and integers.  I added `None` and `slice` because they were an easy way to just tie into Python's existing slicing logic. I don't see `None` and `slice` being used as much in the wild.

We could instead just add a `level_to_slice` function, but I think this is a little more friendly.

Reply:
consider me convinced. plus it's all localized to a short, singular function so it's not that bad re: maintenance. I even thought "what about a function like `level_to_slice`?" but that's effectively what this is 😄